log4j-core/src/main/java/org/apache/logging/log4j/core/impl/JdkMapAdapterStringMap.java

```diff
@@ -47,9 +47,14 @@ public class JdkMapAdapterStringMap implements StringMap {
     private transient String[] sortedKeys;
 
     public JdkMapAdapterStringMap() {
-        this(new HashMap<String, String>());
+        this(new HashMap<String, String>(), false);
     }
 
+    /**
+     * @deprecated for performance reasons since 2.23.
+     *             Use {@link #JdkMapAdapterStringMap(Map, boolean)} instead.
+     */
+    @Deprecated
     public JdkMapAdapterStringMap(final Map<String, String> map) {
         this.map = Objects.requireNonNull(map, "map");
         try {
```

Comment:
I agree that the new ctor explicitly requiring an immutability flag is the way to go forward, though I feel uncomfortable leaving this deprecated ctor as is. This try-catch is the problem, yet we work around it rather than fixing it. Can't we do something else here? For instance,

```
if (knownImmutableMapClasses.contains(map.getClass())) {
    immutable = true;
} else {
    // try-catch
}
```

Reply:
What about:
```java
this(new HashMap<>(map), false);
```
you proposed in #2101?

Now that I removed all references to this constructor in **our** code, I don't have objections about it generating garbage. We could even use:
```java
this(map, true);
```

Reply:
@ppkarwasz, given we are trying to mitigate a performance regression, doesn't `HashMap::new` (or `this(map, true)`) create another by replacing the one from `try-catch`? I mean, are these significantly faster than `try-catch`?

Reply:
Anything is faster than catching an exception. ;-)

I like the `HashMap::new` version better, because it enforces semantics: IMHO a **frozen** `StringMap` should be immutable, while a `StringMap` created from a `Collections.unmodifiableMap` can be modified by changing the underlying map.

As I remarked, I removed all references to the deprecated constructor in our code. On Github I could only find two other projects that use `JdkMapAdapterStringMap`:
- [zenmoto/opentelemetry_logging_prototypes](https://github.com/zenmoto/opentelemetry_logging_prototypes/blob/main/log4j/src/main/java/io/opentelemetry/sdk/proto/LayoutWrapper.java),
- [apple/servicetalk](https://github.com/apple/servicetalk/blob/main/servicetalk-opentracing-log4j2/src/main/java/io/servicetalk/opentracing/log4j2/ServiceTalkTracingThreadContextMap.java).